File: apps/backend/app/schemas/story.py
"""Story Bible / 章节抽取 schemas（Phase 6 — LLM Story Engine）。

这些模型同时用于：LLM 输出解析校验（Pydantic）、API 响应。
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class ExtractedCharacter(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    aliases: list[str] = Field(default_factory=list)
    summary: str = Field(default="", max_length=1000)
    role_hint: str = Field(default="", max_length=50)


class ExtractedLocation(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=1000)


class ExtractedProp(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=1000)


class ExtractedEvent(BaseModel):
    summary: str = Field(min_length=1, max_length=1000)
    importance: Literal["low", "medium", "high"] = "medium"
    characters: list[str] = Field(default_factory=list)


class ChapterExtraction(BaseModel):
    chapter_summary: str = Field(default="", max_length=1000)
    characters: list[ExtractedCharacter] = Field(default_factory=list)
    locations: list[ExtractedLocation] = Field(default_factory=list)
    props: list[ExtractedProp] = Field(default_factory=list)
    events: list[ExtractedEvent] = Field(default_factory=list)


class BibleCharacter(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    aliases: list[str] = Field(default_factory=list)
    summary: str = Field(default="", max_length=2000)
    role_hint: str = Field(default="", max_length=50)


class BibleLocation(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=2000)


class BibleProp(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=2000)


class BibleEvent(BaseModel):
    summary: str = Field(min_length=1, max_length=2000)
    importance: str = "medium"
    characters: list[str] = Field(default_factory=list)
    chapter_index: int = 0


class StoryBible(BaseModel):
    synopsis: str = Field(default="", max_length=5000)
    characters: list[BibleCharacter] = Field(default_factory=list)
    locations: list[BibleLocation] = Field(default_factory=list)
    props: list[BibleProp] = Field(default_factory=list)
    events: list[BibleEvent] = Field(default_factory=list)
    conflicts: list[str] = Field(default_factory=list)
    plotlines: list[str] = Field(default_factory=list)
    foreshadowing: list[str] = Field(default_factory=list)


class AnalysisRequest(BaseModel):
    novel_id: str
    model_id: str
    mode: Literal["full", "merge"] = "full"


class AnalysisJobOut(BaseModel):
    job_id: str
    status: Literal["queued", "running", "completed", "failed", "cancelled"]
    progress: float | None = None
    detail: str = ""
    error: str | None = None
    created_at: datetime


class StoryBibleOut(BaseModel):
    bible: StoryBible | None = None

File: apps/desktop/src/types/story.ts
export type AnalysisMode = "full" | "merge";
export type AnalysisStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export interface BibleCharacter {
  name: string;
  aliases: string[];
  summary: string;
  role_hint: string;
}

export interface BibleLocation {
  name: string;
  description: string;
}

export interface BibleProp {
  name: string;
  description: string;
}

export interface BibleEvent {
  summary: string;
  importance: string;
  characters: string[];
  chapter_index: number;
}

export interface StoryBible {
  synopsis: string;
  characters: BibleCharacter[];
  locations: BibleLocation[];
  props: BibleProp[];
  events: BibleEvent[];
  conflicts: string[];
  plotlines: string[];
  foreshadowing: string[];
}

export interface AnalysisJob {
  job_id: string;
  status: AnalysisStatus;
  progress: number | null;
  detail: string;
  error: string | null;
  created_at: string;
}

export interface AiNovelBrief {
  genre: string;
  audience: string;
  ideas: string;
  complexity: number;
  chapter_count: number;
}

export interface OutlineChapter {
  title: string;
  summary: string;
}

export interface AiOutlineResult {
  title: string;
  chapters: OutlineChapter[];
}

export interface AiChapter {
  title: string;
  content: string;
  summary: string;
}
